File: app/src/main/java/com/stevenfrew/beatprompter/midi/RecursiveAliasComponent.java
package com.stevenfrew.beatprompter.midi;

import com.stevenfrew.beatprompter.BeatPrompterApplication;
import com.stevenfrew.beatprompter.R;

import java.util.List;

public class RecursiveAliasComponent implements AliasComponent {
    private String mReferencedAliasName;
    private List<Value> mArguments;

    public RecursiveAliasComponent(String referencedAliasName,List<Value> arguments)
    {
        mReferencedAliasName=referencedAliasName;
        mArguments=arguments;
    }

    @Override
    public List<OutgoingMessage> resolve(List<Alias> aliases, byte[] parameters, byte channel) throws ResolutionException {
        for(Alias alias:aliases)
            if(alias.mName.equalsIgnoreCase(mReferencedAliasName))
            {
                byte[] newArgs=new byte[mArguments.size()];
                int counter=0;
                for(Value arg:mArguments)
                    newArgs[counter++]=arg.resolve(parameters,channel);
                return alias.resolve(aliases,newArgs,channel);
            }
        throw new ResolutionException(BeatPrompterApplication.getResourceString(R.string.unknown_midi_directive,mReferencedAliasName));
    }
}

File: app/src/main/java/com/stevenfrew/beatprompter/midi/RecursiveAliasComponent.kt
package com.stevenfrew.beatprompter.midi

import com.stevenfrew.beatprompter.BeatPrompterApplication
import com.stevenfrew.beatprompter.R

class RecursiveAliasComponent(private val mReferencedAliasName: String, private val mArguments: List<Value>) : AliasComponent {

    @Throws(ResolutionException::class)
    override fun resolve(aliases: List<Alias>, parameters: ByteArray, channel: Byte): List<OutgoingMessage> {
        try {
            return aliases.first{it.mName.equals(mReferencedAliasName, ignoreCase = true)}.resolve(aliases,mArguments.map{it.resolve(parameters,channel)}.toByteArray(),channel)
        }
        catch(nsee:NoSuchElementException) {
            throw ResolutionException(BeatPrompterApplication.getResourceString(R.string.unknown_midi_directive, mReferencedAliasName))
        }
    }
}
